posts/2022/09-20-pg-stat-statements.md

```diff
@@ -0,0 +1,116 @@
+==title==
+Using pg_stat_statements to find slow queries in Postgres
+
+==author==
+Alan Vardy
+
+==footer==
+
+==description==
+Find slow queries using a built-in Postgres view.
+
+==tags==
+postgres
+
+==body==
+
+![An elephant doing elephant things](elephant.jpg "An elephant doing elephant things")
+[From Unsplash](https://unsplash.com/photos/OP3ghv27Yto)
+
+## Why do we need pg_stat_statements
+
+Slow queries can be one of the worst things to happen to a web application because databases are often the primary source of shared state. Long-running Postgres operations will monopolize connections and tie up the database CPU as it struggles to complete the query. In addition, other queries that would typically not experience issues fail to execute, and the knock-on effect can bring down the whole system.
+
+To determine which queries are slower than others and identify potential problems, we can use a view in Postgres called `pg_stat_statements`. This view is an aggregate of statistics around the queries (and mutations) and helps us identify how much stress these operations put on the database.
+
+We recommend periodically checking the `pg_stat_statements` view for slow queries to catch them before they become a problem, and it is also a way to find them in a hurry. We may even find that optimizing queries can reduce the costs of running the database as fewer resources may need to be assigned to it afterward, or it may not need to have more resources allocated in the future. The database can be the most expensive part of the system to scale; thus, this effort is frequently worth it.
```

Comment:
```suggestion
I recommend periodically checking the `pg_stat_statements` view for slow queries to catch them before they become a problem, and it is also a way to find them in a hurry. We may even find that optimizing queries can reduce the costs of running the database as fewer resources may need to be assigned to it afterward, or it may not need to have more resources allocated in the future. The database can be the part of the system that is most expensive to scale; thus, this effort is frequently worth it.
```